File: R/CARTpscore.R
######################################################################################################################
# Author: Steve Mooney
# Program: CARTpscore.R
# Language: R
# Date: Monday, July 13, 2020
# Data in: DIGdata (asympTest package)
# Description:eAppendix #1: R code that uses CART to compute propensity scores for an 
#   inverse probability weighted estimate the causal effect of Digitalis on death in 
#   the Digitalis Investigator Group Trial
# Released under the GNU General Public License: http://www.gnu.org/copyleft/gpl.html
######################################################################################################################

#install.packages("caret")
#install.packages("asympTest")
# Main code
library(caret)
library(asympTest)
set.seed(12345)
data(DIGdata)

# In the default dataset, the TRTMT variable is coded as 0/1, which makes R think it is a continuous rather than a 
# dichotomous variable.  We recode it here
DIGdata$TRTMT_Dichotomous <- as.factor(DIGdata$TRTMT)
complete_case <- DIGdata[!is.na(DIGdata$BMI),]

# split data
training_dataset <- DIGdata[1:3400,]
test_dataset <- DIGdata[3401:6800,]

# use ten-fold cross validation on the training dataset to pick parameters for a CART model.
control <- trainControl(method="repeatedcv", number=10, repeats=3)
cart_model <- train(TRTMT_Dichotomous~AGE+SEX+BMI, data=complete_case, method="rpart", metric="Accuracy", trControl=control)
print(cart_model)

# Comput propensity scores as probability of treatment (propensity scores)
treatment_prob <- predict(cart_model, complete_case, type="prob")

# Not surprisingly for an RCT, treatment probabilities are decently balanced
plot(density(treatment_prob[,'1']))
lines(density(treatment_prob[,'0']), col='red')

# Use propensity scores for IPW analysis.
weights <- 1/treatment_prob
complete_case$IPW[complete_case$TRTMT_Dichotomous == 1] <- weights[complete_case$TRTMT_Dichotomous == 1, "1"]
complete_case$IPW[complete_case$TRTMT_Dichotomous == 0] <- weights[complete_case$TRTMT_Dichotomous == 0, "0"]
table(is.na(complete_case$IPW))

# Compute an outcome model
outcome_model <- glm(DEATH~TRTMT, data=complete_case, family="binomial", weights=IPW)
all_treated <- complete_case
all_treated$TRTMT <- 1
all_untreated <- complete_case
all_untreated$TRTMT <- 0

# Estimtate outcomes if everyone were treated
predicted_outcome_all_treated <- predict(outcome_model, all_treated, weights=IPW, type='response')
mean(predicted_outcome_all_treated)
predicted_outcome_all_untreated <- predict(outcome_model, all_untreated, weights=IPW, type='response')
mean(predicted_outcome_all_untreated)

# Estimate additive effect
mean(predicted_outcome_all_treated - predicted_outcome_all_untreated)

File: julia/CARTpscore.jl
######################################################################################################################
# Author: Alex Keil
# Program: CARTpscore.jl
# Language: Julia 1.4.2
# Date: Monday, July 13, 2020
# Data in: digdata.csv
# Description: Bonus: Julia code that uses CART to compute propensity scores for 
#   an inverse probability weighted estimate the causal effect of Digitalis on death 
#   in the Digitalis Investigator Group Trial
# Note: if you have not used Julia before, the first time loading a package calls
#  a compiler, which can take a significant amount of time (esp. with plots). The second time
#  running the code will go much more quickly.
# Released under the GNU General Public License: http://www.gnu.org/copyleft/gpl.html
######################################################################################################################

# Run once to install packages
# using Pkg
# Pkg.add(["DecisionTree", "Plots", "CSV", "DataFrames", "Plotly", "CovarianceMatrices", "GLM", "Suppressor"])

using DecisionTree, StatsPlots, GLM, CovarianceMatrices, Statistics, Suppressor
import CSV, DataFrames # similar to "using" but you have to start the function calls with "CSV."
plotly() # use plotly backend for plotting (plotshows up in browser window)

DIGdata = DataFrames.DataFrame(CSV.File("../data/digdata.csv", missingstring="NA"));
cc = DataFrames.dropmissing(DIGdata, :BMI)

# use ten-fold cross validation on the training dataset to pick parameters for a CART model.
y  = string.(cc.TRTMT); # must be string to use classification tree
X = Array(cc[:,[:AGE, :SEX, :BMI]]);

# Julia doesn't use cost-complexity pruning, so we'll optimize on the max tree depth
# will use CV 3 times

println("Cross-validated selection of tree depth")
function get_cvaccuracy(;minTrees = 1, maxTrees = 20, nrepeats=3, K=10)
  # convenience function to calculate 'K' fold cross validated accuracy averaged over 'nrepeats'
  # due to use of nfoldCV_tree function, this will print a lot of info to output
  n_subfeatures=0; 
  min_samples_leaf=1; 
  min_samples_split=2
  min_purity_increase=0.0; 
  pruning_purity = 1.0
  along = collect(minTrees:maxTrees)
  res = zeros(length(along))
  for repeat in 1:nrepeats
    for (j, depth) in enumerate(along)
      foldaccuracy = nfoldCV_tree(y,X, 
                  K,              # number of folds
                  pruning_purity, # pruning_purity (1 = no pruning)
                  depth,          # max depth
                  min_samples_leaf,
                  min_samples_split,
                  min_purity_increase
                )
      res[j] += inv(nrepeats) * mean(foldaccuracy)
    end
  end
  res
end


CVaccuracy = @suppress get_cvaccuracy(minTrees = 1, maxTrees = 20);

# select best performing tree and fit it        
maxdepth = argmin(CVaccuracy)

n_subfeatures=0; 
min_samples_leaf=1; 
min_samples_split=2
min_purity_increase=0.0; 

model = build_tree(y,X,
          n_subfeatures, # number of subfeatures (0 = all)
          maxdepth,
          min_samples_leaf,
          min_samples_split,
          min_purity_increase
        )

# Compute probability of treatment (propensity scores) and non-treatment
probs = apply_tree_proba(model, X, ["1", "0"]);
pscore = probs[:,1];


# plot the propensity scores (will just be peaks)
density(pscore[cc.TRTMT .== 1], label="Treated");
show(density!(pscore[cc.TRTMT .== 0], label="Unreated"))

cc.weights = [x == 1 ? 1.0 / pscore[i] : 1.0 / (1.0 - pscore[i]) for (i,x) in enumerate(cc.TRTMT)];


# fit MSM
msmfit = glm(@formula(DEATH ~ TRTMT), cc, Bernoulli(), wts=cc.weights)
# get robust standard error estimates and construct confidence intervals
beta = coef(msmfit)
robse = stderror(msmfit, HC0())
ci = beta .+ [-1.96 1.96; -1.96 1.96] .* robse

println("odds ratio, 95% robust CI")
println(exp(beta[2]), exp.(ci[2,:]))

# Estimtate outcomes if everyone were treated
cc0 = copy(cc);
cc0.TRTMT .= 0;
pr0 = GLM.predict(msmfit, cc0);
cc1 = copy(cc);
cc1.TRTMT .= 1;
pr1 = GLM.predict(msmfit, cc1);
# Estimate additive effect
println("Predicted outcome, all treated")
println(mean(pr1))
println("Predicted outcome, all untreated")
println(mean(pr0))
println("Additive treatment effect")
println(mean(pr1) - mean(pr0))